# PWD: packages/providers/DownloadProvider
diff --git a/src/com/android/providers/downloads/DownloadProvider.java b/src/com/android/providers/downloads/DownloadProvider.java
index 5f00d49a..109f7751 100644
--- a/src/com/android/providers/downloads/DownloadProvider.java
+++ b/src/com/android/providers/downloads/DownloadProvider.java
@@ -1195,14 +1195,16 @@ public final class DownloadProvider extends ContentProvider {
             if (appOps.noteProxyOp(AppOpsManager.OP_WRITE_EXTERNAL_STORAGE,
                     getCallingPackage()) != AppOpsManager.MODE_ALLOWED) {
                 throw new SecurityException("No permission to write to " + file);
             }
         } else {
             throw new SecurityException("Unsupported path " + file);
         }
+        // check whether record already exists in MP or getCallingPackage owns this file
+        checkWhetherCallingAppHasAccess(file.getPath(), Binder.getCallingUid());
     }
 
     private int getCallingPackageTargetSdkVersion() {
         final String callingPackage = getCallingPackage();
         if (callingPackage != null) {
             ApplicationInfo ai = null;
             try {
@@ -1213,14 +1215,52 @@ public final class DownloadProvider extends ContentProvider {
             if (ai != null) {
                 return ai.targetSdkVersion;
             }
         }
         return Build.VERSION_CODES.CUR_DEVELOPMENT;
     }
 
+    private void checkWhetherCallingAppHasAccess(String filePath, int uid) {
+        try (ContentProviderClient client = getContext().getContentResolver()
+                .acquireContentProviderClient(MediaStore.AUTHORITY)) {
+            if (client == null) {
+                Log.w(Constants.TAG, "Failed to acquire ContentProviderClient for MediaStore");
+                return;
+            }
+
+            final Uri filesUri = MediaStore.setIncludePending(
+                    MediaStore.Files.getContentUriForPath(filePath));
+
+            try (Cursor cursor = client.query(filesUri,
+                    new String[]{MediaStore.Files.FileColumns._ID,
+                            MediaStore.Files.FileColumns.OWNER_PACKAGE_NAME},
+                    MediaStore.Files.FileColumns.DATA + "=?", new String[]{filePath},
+                    null)) {
+                if (cursor != null && cursor.moveToFirst()) {
+                    String fetchedOwnerPackageName = cursor.getString(
+                            cursor.getColumnIndexOrThrow(
+                                    MediaStore.Files.FileColumns.OWNER_PACKAGE_NAME));
+                    String[] packageNames = getContext().getPackageManager().getPackagesForUid(uid);
+
+                    if (fetchedOwnerPackageName != null && packageNames != null) {
+                        boolean isCallerAuthorized = Arrays.asList(packageNames)
+                                .contains(fetchedOwnerPackageName);
+                        if (!isCallerAuthorized) {
+                            throw new SecurityException("Caller does not have access to this path");
+                        }
+                    }
+                }
+            }
+        } catch (RemoteException e) {
+            Log.w(Constants.TAG, "Failed to query MediaStore: " + e.getMessage());
+        }
+    }
+
+
+
     /**
      * Apps with the ACCESS_DOWNLOAD_MANAGER permission can access this provider freely, subject to
      * constraints in the rest of the code. Apps without that may still access this provider through
      * the public API, but additional restrictions are imposed. We check those restrictions here.
      *
      * @param values ContentValues provided to insert()
      * @throws SecurityException if the caller has insufficient permissions
