# PWD: build/make
diff --git a/core/version_defaults.mk b/core/version_defaults.mk
index 65e7f4553c..a3bda42a91 100644
--- a/core/version_defaults.mk
+++ b/core/version_defaults.mk
@@ -248,11 +248,11 @@ ifndef PLATFORM_SECURITY_PATCH
     #  Used to indicate the security patch that has been applied to the device.
     #  It must signify that the build includes all security patches issued up through the designated Android Public Security Bulletin.
     #  It must be of the form "YYYY-MM-DD" on production devices.
     #  It must match one of the Android Security Patch Level strings of the Public Security Bulletins.
     #  If there is no $PLATFORM_SECURITY_PATCH set, keep it empty.
-      PLATFORM_SECURITY_PATCH := 2023-02-05
+      PLATFORM_SECURITY_PATCH := 2023-04-05
 endif
 .KATI_READONLY := PLATFORM_SECURITY_PATCH
 
 ifndef PLATFORM_SECURITY_PATCH_TIMESTAMP
   # Used to indicate the matching timestamp for the security patch string in PLATFORM_SECURITY_PATCH.
