# PWD: packages/services/Telephony
--- a/src/com/android/phone/settings/RadioInfo.java
+++ b/src/com/android/phone/settings/RadioInfo.java
@@ -1740,8 +1740,7 @@ public class RadioInfo extends AppCompatActivity {
         mImsVolteProvisionedSwitch.setOnCheckedChangeListener(null);
         mImsVolteProvisionedSwitch.setChecked(isProvisioned);
         mImsVolteProvisionedSwitch.setOnCheckedChangeListener(mImsVolteCheckedChangeListener);
-        mImsVolteProvisionedSwitch.setEnabled(!IS_USER_BUILD
-                && isEnabledByPlatform && isImsVolteProvisioningRequired());
+        mImsVolteProvisionedSwitch.setEnabled(false);
     }
 
     private void updateVtProvisionedSwitch(boolean isEnabledByPlatform) {
@@ -1762,8 +1761,7 @@ public class RadioInfo extends AppCompatActivity {
         mImsWfcProvisionedSwitch.setOnCheckedChangeListener(null);
         mImsWfcProvisionedSwitch.setChecked(isProvisioned);
         mImsWfcProvisionedSwitch.setOnCheckedChangeListener(mImsWfcCheckedChangeListener);
-        mImsWfcProvisionedSwitch.setEnabled(!IS_USER_BUILD
-                && isEnabledByPlatform && isImsWfcProvisioningRequired());
+        mImsWfcProvisionedSwitch.setEnabled(false);
     }
 
     private void updateEabProvisionedSwitch(boolean isEnabledByPlatform) {

