# PWD: frameworks/base
diff --git a/packages/SettingsLib/src/com/android/settingslib/fuelgauge/PowerAllowlistBackend.java b/packages/SettingsLib/src/com/android/settingslib/fuelgauge/PowerAllowlistBackend.java
index ea8a62f84164..213ab2c6ff2f 100644
--- a/packages/SettingsLib/src/com/android/settingslib/fuelgauge/PowerAllowlistBackend.java
+++ b/packages/SettingsLib/src/com/android/settingslib/fuelgauge/PowerAllowlistBackend.java
@@ -72,15 +72,7 @@ public class PowerAllowlistBackend {
     }
 
     public boolean isAllowlisted(String pkg) {
-        if (mAllowlistedApps.contains(pkg)) {
-            return true;
-        }
-
-        if (isDefaultActiveApp(pkg)) {
-            return true;
-        }
-
-        return false;
+        return mAllowlistedApps.contains(pkg);
     }
 
     /**
