# PWD: packages/apps/Settings

diff --git a/src/com/android/settings/development/AdbRootPreferenceController.java b/src/com/android/settings/development/AdbRootPreferenceController.java
index 5f7d65da35..8cdc24ff8a 100644
--- a/src/com/android/settings/development/AdbRootPreferenceController.java
+++ b/src/com/android/settings/development/AdbRootPreferenceController.java
@@ -53,7 +53,7 @@ public class AdbRootPreferenceController extends DeveloperOptionsPreferenceContr
 
     @Override
     public boolean isAvailable() {
-        return Build.IS_DEBUGGABLE;
+        return true;
     }
 
     @Override
