# PWD: vendor/nxp/opensource/commonsys/packages/apps/Nfc
diff --git a/src/com/android/nfc/NfcService.java b/src/com/android/nfc/NfcService.java
index 0f12d2e9..c69c9466 100644
--- a/src/com/android/nfc/NfcService.java
+++ b/src/com/android/nfc/NfcService.java
@@ -1130,14 +1130,18 @@ public class NfcService implements DeviceHostListener {
                             PackageManager.DONT_KILL_APP,
                     uh.getIdentifier());
         } catch (RemoteException e) {
             Log.w(TAG, "Unable to change Beam status for user " + uh);
         }
     }
 
+    public boolean isSecureNfcEnabled() {
+        return mIsSecureNfcEnabled;
+    }
+
     final class NfcAdapterService extends INfcAdapter.Stub {
         @Override
         public boolean enable() throws RemoteException {
             synchronized (NfcService.this) {
                 if (mNxpNfcState != NXP_NFC_STATE_OFF) {
                     Log.e(TAG, "mNxpNfcStateis not equal to NXP_NFC_STATE_OFF."
                                 + " Enable NFC Rejected.");
diff --git a/src/com/android/nfc/cardemulation/HostEmulationManager.java b/src/com/android/nfc/cardemulation/HostEmulationManager.java
index 6af4e0d7..b2670ec2 100644
--- a/src/com/android/nfc/cardemulation/HostEmulationManager.java
+++ b/src/com/android/nfc/cardemulation/HostEmulationManager.java
@@ -173,16 +173,17 @@ public class HostEmulationManager {
                     return;
                 }
                 mLastSelectedAid = selectAid;
                 if (resolveInfo.defaultService != null) {
                     // Resolve to default
                     // Check if resolvedService requires unlock
                     NfcApduServiceInfo defaultServiceInfo = resolveInfo.defaultService;
-                    if (defaultServiceInfo.requiresUnlock() &&
-                            mKeyguard.isKeyguardLocked() && mKeyguard.isKeyguardSecure()) {
+                    if ((defaultServiceInfo.requiresUnlock()
+                            || NfcService.getInstance().isSecureNfcEnabled())
+                          && mKeyguard.isKeyguardLocked() && mKeyguard.isKeyguardSecure()) {
                         // Just ignore all future APDUs until next tap
                         mState = STATE_W4_DEACTIVATE;
                         launchTapAgain(resolveInfo.defaultService, resolveInfo.category);
                         return;
                     }
                     // In no circumstance should this be an OffHostService -
                     // we should never get this AID on the host in the first place
