# PWD: packages/providers/MediaProvider
diff --git a/jni/FuseDaemon.cpp b/jni/FuseDaemon.cpp
index f71140546..232682d72 100755
--- a/jni/FuseDaemon.cpp
+++ b/jni/FuseDaemon.cpp
@@ -98,15 +98,17 @@ class ScopedTrace {
 const bool IS_OS_DEBUGABLE = android::base::GetIntProperty("ro.debuggable", 0);
 
 #define FUSE_UNKNOWN_INO 0xffffffff
 
 // Stolen from: android_filesystem_config.h
 #define AID_APP_START 10000
 
-constexpr size_t MAX_READ_SIZE = 128 * 1024;
+#define FUSE_MAX_MAX_PAGES 256
+
+const size_t MAX_READ_SIZE = FUSE_MAX_MAX_PAGES * getpagesize();
 // Stolen from: UserHandle#getUserId
 constexpr int PER_USER_RANGE = 100000;
 
 // Stolen from: UserManagerService
 constexpr int MAX_USER_ID = UINT32_MAX / PER_USER_RANGE;
 
 const int MY_UID = getuid();
