# PWD: build/make
diff --git a/core/version_defaults.mk b/core/version_defaults.mk
index 65e7f4553c..1f3a266ab6 100644
--- a/core/version_defaults.mk
+++ b/core/version_defaults.mk
@@ -250,7 +250,7 @@ ifndef PLATFORM_SECURITY_PATCH
     #  It must be of the form "YYYY-MM-DD" on production devices.
     #  It must match one of the Android Security Patch Level strings of the Public Security Bulletins.
     #  If there is no $PLATFORM_SECURITY_PATCH set, keep it empty.
-      PLATFORM_SECURITY_PATCH := 2023-02-05
+      PLATFORM_SECURITY_PATCH := 2023-03-05
 endif
 .KATI_READONLY := PLATFORM_SECURITY_PATCH
 
