# PWD: system/bt
diff --git a/bta/av/bta_av_aact.cc b/bta/av/bta_av_aact.cc
index 33298d64977..73eb8870933 100644
--- a/bta/av/bta_av_aact.cc
+++ b/bta/av/bta_av_aact.cc
@@ -1740,15 +1740,15 @@ void bta_av_setconfig_rej(tBTA_AV_SCB* p_scb, tBTA_AV_DATA* p_data) {
   uint8_t avdt_handle = p_data->ci_setconfig.avdt_handle;
 
   bta_av_adjust_seps_idx(p_scb, avdt_handle);
   LOG_INFO("%s: sep_idx=%d avdt_handle=%d bta_handle=0x%x", __func__,
            p_scb->sep_idx, p_scb->avdt_handle, p_scb->hndl);
   AVDT_ConfigRsp(p_scb->avdt_handle, p_scb->avdt_label, AVDT_ERR_UNSUP_CFG, 0);
 
-  reject.bd_addr = p_data->str_msg.bd_addr;
+  reject.bd_addr = p_scb->PeerAddress();
   reject.hndl = p_scb->hndl;
 
   tBTA_AV bta_av_data;
   bta_av_data.reject = reject;
   (*bta_av_cb.p_cback)(BTA_AV_REJECT_EVT, &bta_av_data);
 }
 
diff --git a/stack/avdt/avdt_msg.cc b/stack/avdt/avdt_msg.cc
index 5e08d02872b..faf9fe5b9cb 100644
--- a/stack/avdt/avdt_msg.cc
+++ b/stack/avdt/avdt_msg.cc
@@ -1581,14 +1581,22 @@ void avdt_msg_ind(AvdtpCcb* p_ccb, BT_HDR* p_buf) {
       ok = false;
 
       /* send a general reject */
       if (msg_type == AVDT_MSG_TYPE_CMD) {
         avdt_msg_send_grej(p_ccb, sig, &msg);
       }
     }
+
+    /* validate reject/response against cached sig */
+    if (((msg_type == AVDT_MSG_TYPE_RSP) || (msg_type == AVDT_MSG_TYPE_REJ)) &&
+        (p_ccb->p_curr_cmd == nullptr || p_ccb->p_curr_cmd->event != sig)) {
+      AVDT_TRACE_WARNING(
+          "Dropping msg with mismatched sig; sig=%d", sig);
+      ok = false;
+    }
   }
 
   if (ok && !gen_rej) {
     /* skip over header (msg length already verified during reassembly) */
     p_buf->len -= AVDT_LEN_TYPE_SINGLE;
 
     /* set up to parse message */
diff --git a/stack/btm/btm_ble.cc b/stack/btm/btm_ble.cc
index 601d85be015..6542e46303d 100644
--- a/stack/btm/btm_ble.cc
+++ b/stack/btm/btm_ble.cc
@@ -1376,14 +1376,17 @@ tBTM_STATUS btm_ble_set_encryption(const RawAddress& bd_addr,
   if (sec_act == BTM_BLE_SEC_ENCRYPT_MITM) {
     p_rec->security_required |= BTM_SEC_IN_MITM;
   }
 
   switch (sec_act) {
     case BTM_BLE_SEC_ENCRYPT:
       if (link_role == HCI_ROLE_CENTRAL) {
+        if (p_rec->is_le_device_encrypted()) {
+          return BTM_SUCCESS;
+        }
         /* start link layer encryption using the security info stored */
         cmd = btm_ble_start_encrypt(bd_addr, false, NULL);
         break;
       }
       /* if salve role then fall through to call SMP_Pair below which will send
          a sec_request to request the central to encrypt the link */
       FALLTHROUGH_INTENDED; /* FALLTHROUGH */
@@ -1720,14 +1723,43 @@ void btm_ble_connected(const RawAddress& bda, uint16_t handle, uint8_t enc_mode,
 void btm_ble_connected_from_address_with_type(
     const tBLE_BD_ADDR& address_with_type, uint16_t handle, uint8_t enc_mode,
     uint8_t role, bool addr_matched) {
   btm_ble_connected(address_with_type.bda, handle, enc_mode, role,
                     address_with_type.type, addr_matched);
 }
 
+static bool btm_ble_complete_evt_ignore(const tBTM_SEC_DEV_REC* p_dev_rec,
+                                        const tSMP_EVT_DATA* p_data) {
+  // Encryption request in peripheral role results in SMP Security request. SMP
+  // may generate a SMP_COMPLT_EVT failure event cases like below: 1) Some
+  // central devices don't handle cross-over between encryption and SMP security
+  // request 2) Link may get disconnected after the SMP security request was
+  // sent.
+  if (p_data->cmplt.reason != SMP_SUCCESS && !p_dev_rec->role_central &&
+      btm_cb.pairing_bda != p_dev_rec->bd_addr &&
+      btm_cb.pairing_bda != p_dev_rec->ble.pseudo_addr &&
+      p_dev_rec->is_le_link_key_known() &&
+      p_dev_rec->ble.key_type != BTM_LE_KEY_NONE) {
+    if (p_dev_rec->is_le_device_encrypted()) {
+      LOG_WARN("Bonded device %s is already encrypted, ignoring SMP failure",
+               PRIVATE_ADDRESS(p_dev_rec->bd_addr));
+      return true;
+    } else if (p_data->cmplt.reason == SMP_CONN_TOUT) {
+      LOG_WARN(
+          "Bonded device %s disconnected while waiting for encryption, "
+          "ignoring SMP failure",
+          PRIVATE_ADDRESS(p_dev_rec->bd_addr));
+      l2cu_start_post_bond_timer(p_dev_rec->ble_hci_handle);
+      return true;
+    }
+  }
+
+  return false;
+}
+
 /*****************************************************************************
  *  Function        btm_proc_smp_cback
  *
  *  Description     This function is the SMP callback handler.
  *
  *****************************************************************************/
 tBTM_STATUS btm_proc_smp_cback(tSMP_EVT event, const RawAddress& bd_addr,
@@ -1767,14 +1799,18 @@ tBTM_STATUS btm_proc_smp_cback(tSMP_EVT event, const RawAddress& bd_addr,
         if (event != SMP_CONSENT_REQ_EVT) {
           p_dev_rec->sec_state = BTM_SEC_STATE_AUTHENTICATING;
         }
         btm_cb.pairing_flags |= BTM_PAIR_FLAGS_LE_ACTIVE;
         FALLTHROUGH_INTENDED; /* FALLTHROUGH */
 
       case SMP_COMPLT_EVT:
+        if (btm_ble_complete_evt_ignore(p_dev_rec, p_data)) {
+          return BTM_SUCCESS;
+        }
+
         if (btm_cb.api.p_le_callback) {
           /* the callback function implementation may change the IO
            * capability... */
           BTM_TRACE_DEBUG("btm_cb.api.p_le_callback=0x%x",
                           btm_cb.api.p_le_callback);
           (*btm_cb.api.p_le_callback)(event, bd_addr,
                                       (tBTM_LE_EVT_DATA*)p_data);
diff --git a/stack/l2cap/l2c_ble.cc b/stack/l2cap/l2c_ble.cc
index c878ef2a543..245cdb89d04 100644
--- a/stack/l2cap/l2c_ble.cc
+++ b/stack/l2cap/l2c_ble.cc
@@ -188,18 +188,18 @@ void l2cble_notify_le_connection(const RawAddress& bda) {
     LOG_WARN("Received notification for le connection but no lcb found");
     return;
   }
 
   if (BTM_IsAclConnectionUp(bda, BT_TRANSPORT_LE) &&
       p_lcb->link_state != LST_CONNECTED) {
     /* update link status */
+    p_lcb->link_state = LST_CONNECTED;
     // TODO Move this back into acl layer
     btm_establish_continue_from_address(bda, BT_TRANSPORT_LE);
     /* update l2cap link status and send callback */
-    p_lcb->link_state = LST_CONNECTED;
     l2cu_process_fixed_chnl_resp(p_lcb);
   }
 
   /* For all channels, send the event through their FSMs */
   for (tL2C_CCB* p_ccb = p_lcb->ccb_queue.p_first_ccb; p_ccb;
        p_ccb = p_ccb->p_next_ccb) {
     if (p_ccb->chnl_state == CST_CLOSED)
diff --git a/stack/sdp/sdp_discovery.cc b/stack/sdp/sdp_discovery.cc
index 72f2413a0fb..e82c2900683 100644
--- a/stack/sdp/sdp_discovery.cc
+++ b/stack/sdp/sdp_discovery.cc
@@ -844,14 +844,19 @@ static uint8_t* add_attr(uint8_t* p, uint8_t* p_end, tSDP_DISCOVERY_DB* p_db,
   uint16_t id;
   uint8_t type;
   uint8_t* p_attr_end;
   uint8_t is_additional_list = nest_level & SDP_ADDITIONAL_LIST_MASK;
 
   nest_level &= ~(SDP_ADDITIONAL_LIST_MASK);
 
+  if (p + sizeof(uint8_t) > p_end) {
+    SDP_TRACE_WARNING("bad arguments to add_addr", __func__);
+    return NULL;
+  }
+
   type = *p++;
   p = sdpu_get_len_from_type(p, p_end, type, &attr_len);
   if (p == NULL || (p + attr_len) > p_end) {
     SDP_TRACE_WARNING("%s: bad length in attr_rsp", __func__);
     return NULL;
   }
   attr_len &= SDP_DISC_ATTR_LEN_MASK;
diff --git a/stack/smp/smp_act.cc b/stack/smp/smp_act.cc
index fcdc7a88c99..a80396f4e94 100644
--- a/stack/smp/smp_act.cc
+++ b/stack/smp/smp_act.cc
@@ -1930,14 +1930,24 @@ void smp_process_secure_connection_oob_data(tSMP_CB* p_cb,
   if (p_sc_oob_data->loc_oob_data.present) {
     p_cb->local_random = p_sc_oob_data->loc_oob_data.randomizer;
   } else {
     SMP_TRACE_EVENT("%s: local OOB randomizer is absent", __func__);
     p_cb->local_random = {0};
   }
 
+  if (p_cb->peer_oob_flag == SMP_OOB_PRESENT && !p_sc_oob_data->loc_oob_data.present) {
+    SMP_TRACE_WARNING(
+        "local OOB data is not present but peer claims to have received it; dropping "
+        "connection", __func__);
+    tSMP_INT_DATA smp_int_data{};
+    smp_int_data.status = SMP_OOB_FAIL;
+    smp_sm_event(p_cb, SMP_AUTH_CMPL_EVT, &smp_int_data);
+    return;
+  }
+
   if (!p_sc_oob_data->peer_oob_data.present) {
     SMP_TRACE_EVENT("%s: peer OOB data is absent", __func__);
     p_cb->peer_random = {0};
   } else {
     p_cb->peer_random = p_sc_oob_data->peer_oob_data.randomizer;
     p_cb->remote_commitment = p_sc_oob_data->peer_oob_data.commitment;
 
